# **SQL Project**  

The COVID-19 pandemic took the world by surprise, disrupting daily life. For a time, people stopped going out, visiting cafes, and shopping malls. However, this created more time for reading books. Entrepreneurs quickly noticed this trend and started developing apps for book lovers.  

Your company decided to stay ahead of the curve and acquired a major subscription-based book reading service. As an analyst, your first task is to explore its database.  

The database contains information about books, publishers, authors, and user reviews. These insights will help shape the value proposition of the new product.  

---

**Data Description**  

**Table: books**  
Contains information about books:  
- **book_id** – unique identifier of the book;  
- **author_id** – unique identifier of the author;  
- **title** – book title;  
- **num_pages** – number of pages;  
- **publication_date** – book publication date;  
- **publisher_id** – unique identifier of the publisher.  

**Table: authors**  
Contains information about authors:  
- **author_id** – unique identifier of the author;  
- **author** – author's name.  

**Table: publishers**  
Contains information about publishers:  
- **publisher_id** – unique identifier of the publisher;  
- **publisher** – publisher's name.  

**Table: ratings**  
Contains user rating data:  
- **rating_id** – unique identifier of the rating;  
- **book_id** – unique identifier of the book;  
- **username** – name of the user who left the rating;  
- **rating** – book rating.  

**Table: reviews**  
Contains user review data:  
- **review_id** – unique identifier of the review;  
- **book_id** – unique identifier of the book;  
- **username** – name of the user who wrote the review;  
- **text** – review content.  

---

**Tasks**  

- Count the number of books published after January 1, 2000.  

- Calculate the number of reviews and the average rating for each book.  

- Identify the publisher that released the highest number of books with more than 50 pages (excluding pamphlets).  

- Determine the author with the highest average book rating—consider only books with at least 50 ratings.  

- Calculate the average number of reviews from users who have given more than 48 ratings.  

## Connecting to the Database

In [1]:
# Importing libraries
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# Setting parameters
db_config = {
    'user': 'praktikum_student',  # username
    'pwd': 'Sdf4$2;d-d30pp',  # password
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432,  # connection port
    'db': 'data-analyst-final-project-db'  # database name
}

connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# Saving the connector
engine = create_engine(connection_string, connect_args={'sslmode': 'require'})

con = engine.connect()

## Exploring the Data
We will examine the tables by displaying the first few rows.

In [3]:
# Function to Retrieve the First Rows of a Table
def preview_table(table_name, con, limit=5):
    query = f'SELECT * FROM {table_name} LIMIT {limit}'
    df = pd.io.sql.read_sql(sql=text(query), con=con)
    print(f'\n{table_name.upper()}')
    display(df)
    
    # Getting information about the entire table
    full_query = f'SELECT * FROM {table_name}'
    full_df = pd.io.sql.read_sql(sql=text(full_query), con=con)
    display(full_df.info())

In [4]:
# List of tables for the study
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

# Scroll through the list of tables and display the first rows
for table in tables:
    preview_table(table, con)


BOOKS


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None


AUTHORS


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None


PUBLISHERS


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None


RATINGS


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None


REVIEWS


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

## Executing Project Tasks  

### Count the Number of Books Published After January 1, 2000

In [5]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

In [6]:
query_count = '''SELECT COUNT(DISTINCT book_id) FROM books 
WHERE CAST(publication_date AS date) >= '2000-01-01';'''
select(query_count)

count
0    821

821 books have been published since January 1, 2000.

### For each book, count the number of reviews and the average rating

In [7]:
query_rating_review = '''
SELECT 
    b.book_id, 
    b.title, 
    COUNT(DISTINCT r.review_id) AS review_count, 
    ROUND(AVG(rt.rating), 2) AS avg_rating
FROM books b
LEFT JOIN reviews r ON b.book_id = r.book_id
LEFT JOIN ratings rt ON b.book_id = rt.book_id
GROUP BY b.book_id, b.title
ORDER BY review_count DESC;
'''

print("\nThe number of reviews and the average score for each book:")
select(query_rating_review)


The number of reviews and the average score for each book:


book_id                                              title  review_count  \
0        948                            Twilight (Twilight  #1)             7   
1        963                                Water for Elephants             6   
2        734                                   The Glass Castle             6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
4        695  The Curious Incident of the Dog in the Night-Time             6   
..       ...                                                ...           ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
996      808                            The Natural Way to Draw             0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998      221                          Essential Tales and Poems             0   
999      191  Disney's Beauty and the Beast (A Little Golden...             0   

     avg_rating  
0          3.66  
1          3.98  
2          4.21  
3          4.41  
4          4.08  
..          ...  
995        3.67  
996        3.00  
997        5.00  
998        4.00  
999        4.00  

[1000 rows x 4 columns]

### Identify the Publisher That Released the Most Books Over 50 Pages – Excluding Pamphlets

In [8]:
query_publisher = '''
SELECT 
    p.publisher_id, 
    p.publisher, 
    COUNT(b.book_id) AS book_count
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id, p.publisher
ORDER BY book_count DESC
LIMIT 1;
'''
select(query_publisher)

publisher_id      publisher  book_count
0           212  Penguin Books          42

The publisher that has released the largest number of books thicker than 50 pages: **Penguin Books**

### Identify the Author with the Highest Average Book Rating – Considering Only Books with 50 or More Ratings

In [9]:
query_author = '''
WITH book_ratings AS (
    SELECT 
        b.book_id,
        b.author_id,
        COUNT(rt.rating) AS rating_count,
        AVG(rt.rating) AS avg_rating
    FROM books b
    JOIN ratings rt ON b.book_id = rt.book_id
    GROUP BY b.book_id, b.author_id
    HAVING COUNT(rt.rating) >= 50
),
author_avg_rating AS (
    SELECT 
        br.author_id,
        ROUND(AVG(br.avg_rating), 2) AS avg_rating,
        SUM(br.rating_count) AS total_ratings
    FROM book_ratings br
    GROUP BY br.author_id
)
SELECT 
    a.author,
    ar.avg_rating
FROM author_avg_rating ar
JOIN authors a ON ar.author_id = a.author_id
ORDER BY ar.avg_rating DESC
LIMIT 1;
'''

select(query_author)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.28

The author with the highest average book rating (considering books with 50 or more ratings) is **J.K. Rowling/Mary GrandPré** (average rating: 4.28).

### Calculate the Average Number of Reviews from Users Who Gave More Than 48 Ratings

In [10]:
query_avg_reviews = '''
WITH user_reviews AS (
    SELECT 
        rv.username, 
        COUNT(rv.review_id) AS review_count
    FROM reviews rv
    JOIN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating) > 48
    ) AS frequent_users ON rv.username = frequent_users.username
    GROUP BY rv.username
)
SELECT 
    AVG(review_count) AS avg_reviews
FROM user_reviews;
'''


select(query_avg_reviews)

avg_reviews
0         24.0

The average number of reviews from users who gave more than 48 ratings is **24**.

In [11]:
# Closing the сonnection
con.close()

## **Conclusions**  

- The dataset contains information on **1,000 books**.  

- **821 books** were published after January 1, 2000.  

- **Number of reviews and average rating for each book:**  
  The book *"Twilight (Twilight #1)"* has the highest number of reviews (**7**) with an average rating of **3.66**. Other books, including *"Water for Elephants"* and *"The Glass Castle"*, have **6 reviews** each with ratings above **4.0**.  

- **The publisher that released the most books over 50 pages:** *Penguin Books*.  

- **The author with the highest average book rating (considering books with 50 or more ratings):** *J.K. Rowling/Mary GrandPré* (average rating: **4.28**).  

- **The average number of reviews from users who gave more than 48 ratings:** **24**.  